File: LC3 - Longest Substring without Repeating Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """
        Can't discard entire substring when duplicate is found.
        Need a sliding window, ie Two Pointer approach.

        Keep dictionary of seen chars with last pos, if duplicate found,
        jump to dict[suplicate_char] -> {pos} + 1 and keep track of length
        using left + right pointers.
        """

        left: int = 0
        seen: dict = {}
        max_len: int = 0
        
        for right, char in enumerate(s):
            if char in seen and seen[char] >= left:
                # exclude duplicate in O(1)
                left = seen[char] + 1
                
            # update last seen dict
            seen[char] =  right
            # update max
            max_len = max(max_len, right - left + 1)
         
        return max_len

File: LC3 - Longest Substring without Repeating Characters/Solution.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        // track last seen pos of chars: char -> last index
        HashMap<Character, Integer> seen = new HashMap<>();
        // dual pointers for sliding window
        int left = 0, maxLength = 0;
        
        for (int r=0; r < s.length(); r++){
            int lastPos = seen.getOrDefault(s.charAt(r),-1);
            if (lastPos >= left){
                // duplicate in window, move left ptr past duplicate
                left = lastPos + 1;
            }
            seen.put(s.charAt(r), r);
            int len = r - left + 1;
            maxLength = Math.max(len, maxLength);
        }
        return maxLength;
    }
}